src/uu/pgrep/src/pgrep.rs

```diff
@@ -0,0 +1,321 @@
+        .get_many::<String>("pattern")
+        .unwrap_or_default()
+        .map(|it| it.to_string());
+
+    if should_ignore_case {
+        patterns.map(|it| it.to_lowercase()).collect::<Vec<_>>()
+    } else {
+        patterns.collect()
+    }
+}
+
+/// Collect pids with filter construct from command line arguments
+fn collect_matched_pids(matches: &ArgMatches) -> Vec<ProcessInformation> {
+    let binding = String::from("");
+    let pattern = matches.get_one::<String>("pattern").unwrap_or(&binding);
+
+    let should_inverse = matches.get_flag("inverse");
+    let should_ignore_case = matches.get_flag("ignore-case");
+
+    let flag_full = matches.get_flag("full");
+    let flag_exact = matches.get_flag("exact");
+
+    let evaluate = |mut pid: ProcessInformation| {
+        let binding = pid.status();
+        let name = binding.get("Name")?;
+
+        // Process flag `--ignore-case`
+        let (name, pattern) = if should_ignore_case {
+            (name.to_lowercase(), pattern.to_lowercase())
+        } else {
+            (name.into(), pattern.into())
+        };
+
+        // Process flag `--full` && `--exact`
+        let name_matched = if flag_exact {
+            // Equals `Name` in /proc/<pid>/status
+            // The `unwrap` operation must succeed
+            // because the REGEX has been verified as correct in `uumain`.
+            REGEX.get().unwrap().is_match(&name)
+        } else if flag_full {
+            // Equals `cmdline` in /proc/<pid>/cmdline
+            pid.cmdline.contains(&pattern)
+        } else {
+            name.contains(&pattern)
+        };
+
+        // Process flag `--terminal`
+        let tty_matched = if let Some(ttys) = matches.get_many::<String>("terminal") {
+            // convert from input like `pts/0`
+            let ttys = ttys
+                .cloned()
+                .flat_map(TerminalType::try_from)
+                .collect::<HashSet<_>>();
+
+            if let Ok(value) = pid.ttys() {
+                value.iter().any(|it| ttys.contains(it))
+            } else {
+                false
+            }
+        } else {
+            true
+        };
+
+        // Process arg `--runstates`
+        // After testing, I found that GNU implementation
+        // just matching the char in the argument string.
+        let runstate_matched = if let (Some(arg_runstates), Ok(pid_state)) =
+            (matches.get_one::<String>("runstates"), pid.run_state())
+        {
+            arg_runstates.contains(&pid_state.to_string())
+        } else {
+            true
+        };
+
+        if (name_matched && tty_matched && runstate_matched) ^ should_inverse {
+            Some(pid)
+        } else {
+            None
+        }
+    };
+
+    let mut result = Vec::new();
+
+    for pid in walk_process() {
+        if let Some(pid) = evaluate(pid) {
+            result.push(pid.clone())
+        }
+    }
+
+    result
+}
+
+// Make -o and -n as a group of args
+fn filter_oldest_newest(
```

Comment:
I think this function is unnecessary. The list of PIDs you get from `collect_matched_pids` seems to be sorted (I'm not sure whether this is always the case) and so you can simply call `.first()`/`.last()` to get the oldest/newest PIDs.